src/repo/authRepo/getClient/mongo.ts

```
@@ -1,25 +1,56 @@
 import * as atob from 'atob';
-import { ObjectID } from 'mongodb';
+import { Db } from 'mongodb';
 import NoModel from 'jscommons/dist/errors/NoModel';
 import ClientModel from '../../../models/ClientModel';
-import Signature, { Opts, Result } from './Signature';
+import Signature from './Signature';
 import Actor from '../../../models/Actor';
 import FacadeConfig from '../utils/mongoAuth/FacadeConfig';
 import parseJson from '../../../utils/parseJson';
 import ExpiredClientError from '../../../errors/ExpiredClientError';
 import UntrustedClientError from '../../../errors/UntrustedClientError';
 
+type FindClientWithAuth = (db: Db) => (token: string) => Promise<any>;
+
+const findClientByAccessToken: FindClientWithAuth = (db) => async (accessToken) => {
+  const accessTokenDoc = await db.collection('oAuthTokens').findOne({
+    accessToken,
+  });
+  if (!accessTokenDoc) {
+    return null;
+  }
+  const clientDoc = await db.collection('client').findOne({
+    _id: accessTokenDoc.clientId,
+  });
+  return clientDoc;
+};
+
+const findClientByBasicAuth: FindClientWithAuth = (db) => async (encodedBasicAuthToken) => {
+  const decodedAuthToken = atob(encodedBasicAuthToken);
+  const splitAuthToken = decodedAuthToken.split(':');
+  const [key, secret] = splitAuthToken;
+  const clientDoc = await db.collection('client').findOne({
+    'api.basic_key': key,
+    'api.basic_secret': secret,
+  });
+  return clientDoc;
+};
+
+const findClientWithAuth = (authType: string): FindClientWithAuth => {
+  switch (authType) {
+    case 'Basic':
+      return findClientByBasicAuth;
+    case 'Bearer':
+      return findClientByAccessToken;
+    default:
+      return () => async () => null;
```

I think it would be easier to read if we use `throw new NoModel('Client');` error instead of returning an error value like null in these functions, otherwise you have to read through the rest of the code to realise that this just throws an error in the end.

Might be a good idea to extract the `throw new NoModel('Client');` into its own function in this file to make it easier to reuse and modify.

---

Sure. I modified this code. Can you review updated version?
https://github.com/LearningLocker/xapi-statements/pull/339/commits/4c063081ae89c0aeb242883aceee77018daf2dae

---

Yep reviewed it. Good idea to wait until this was reviewed before changing the others.